File: src/WordPattern.java
// Check word pattern.
// Time Complexity - O(n)
// Space Complexity - O(1) ----> In worst case, we only have to store 26 characters
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

class WordPattern {
    public boolean wordPattern(String pattern, String s) {
        Map<Character,String> sMap = new HashMap<>();
        HashSet<String> tSet = new HashSet<>();
        String words[] = s.split(" ");
        if(pattern.length() != words.length) return false;
        for(int i=0;i<pattern.length();i++) {
            char sChar = pattern.charAt(i);
            String tChar = words[i];
            if(sMap.containsKey(sChar)) {
                String val = sMap.get(sChar);
                if(!val.equals(tChar)) return false;
            } else {
                if(!tSet.contains(tChar)) {
                    sMap.put(sChar,tChar);
                    tSet.add(tChar);
                } else{
                    return false;
                }
            }
        }
        return true;
    }
}

File: WordPattern.js
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
var wordPattern = function(pattern, s) {
    var sArray = s.split(' ');
    var sl = sArray.length;
    var pl = pattern.length;
    if(sl != pl) {
        console.log(`sl ${sl}, pl ${pl}`);
        return false;
    }
    var map = {};
    var sMap = {};
    for(let i =0; i< pl; i++){
        
        if( map.hasOwnProperty(pattern[i])){
            if(map[pattern[i]] != sArray[i]) return false;
        }else map[pattern[i]] = sArray[i];
        if( sMap.hasOwnProperty(sArray[i])){
            if(sMap[sArray[i]] != pattern[i]) return false;
        }else sMap[sArray[i]] = pattern[i];
    }
    return true;
};